File: internal/tui/components/toast.go
package components

import (
	"time"

	"github.com/charmbracelet/lipgloss"

	"ash/internal/tui"
)

// ToastType defines the type of toast message.
type ToastType int

const (
	ToastInfo ToastType = iota
	ToastSuccess
	ToastWarning
	ToastError
)

// Toast represents a temporary status message.
type Toast struct {
	styles    tui.Styles
	message   string
	toastType ToastType
	visible   bool
	expiresAt time.Time
	duration  time.Duration
}

// NewToast creates a new toast component.
func NewToast(styles tui.Styles) *Toast {
	return &Toast{
		styles:   styles,
		duration: 3 * time.Second,
	}
}

// Show displays a toast message.
func (t *Toast) Show(message string, toastType ToastType) {
	t.message = message
	t.toastType = toastType
	t.visible = true
	t.expiresAt = time.Now().Add(t.duration)
}

// ShowInfo displays an info toast.
func (t *Toast) ShowInfo(message string) {
	t.Show(message, ToastInfo)
}

// ShowSuccess displays a success toast.
func (t *Toast) ShowSuccess(message string) {
	t.Show(message, ToastSuccess)
}

// ShowWarning displays a warning toast.
func (t *Toast) ShowWarning(message string) {
	t.Show(message, ToastWarning)
}

// ShowError displays an error toast.
func (t *Toast) ShowError(message string) {
	t.Show(message, ToastError)
}

// Hide hides the toast.
func (t *Toast) Hide() {
	t.visible = false
}

// SetDuration sets the toast duration.
func (t *Toast) SetDuration(d time.Duration) {
	t.duration = d
}

// IsVisible returns whether the toast is visible.
func (t *Toast) IsVisible() bool {
	if !t.visible {
		return false
	}

	// Check if toast has expired
	if time.Now().After(t.expiresAt) {
		t.visible = false
		return false
	}

	return true
}

// Render renders the toast.
func (t *Toast) Render() string {
	if !t.IsVisible() {
		return ""
	}

	var style lipgloss.Style
	var icon string

	switch t.toastType {
	case ToastInfo:
		style = lipgloss.NewStyle().
			Foreground(lipgloss.Color("#fafafa")).
			Background(lipgloss.Color("#404040"))
		icon = "i"
	case ToastSuccess:
		style = lipgloss.NewStyle().
			Foreground(lipgloss.Color("#fafafa")).
			Background(lipgloss.Color("#22c55e"))
		icon = "+"
	case ToastWarning:
		style = lipgloss.NewStyle().
			Foreground(lipgloss.Color("#0a0a0a")).
			Background(lipgloss.Color("#f59e0b"))
		icon = "!"
	case ToastError:
		style = lipgloss.NewStyle().
			Foreground(lipgloss.Color("#fafafa")).
			Background(lipgloss.Color("#ef4444"))
		icon = "x"
	}

	style = style.Padding(0, 2)

	return style.Render(icon + " " + t.message)
}

// RenderWithWidth renders the toast with a specific width.
func (t *Toast) RenderWithWidth(width int) string {
	if !t.IsVisible() {
		return ""
	}

	rendered := t.Render()

	return lipgloss.NewStyle().
		Width(width).
		Align(lipgloss.Center).
		Render(rendered)
}

File: ash-zig/src/tui/components/toast.zig
const std = @import("std");
const ansi = @import("../ansi.zig");
const theme = @import("../theme.zig");

pub const ToastType = enum {
    info,
    success,
    warning,
    err,
};

pub const Toast = struct {
    message: []const u8 = "",
    toast_type: ToastType = .info,
    visible: bool = false,
    expires_at: i64 = 0,
    duration_ms: i64 = 3000,
    style: theme.Style = theme.Style.init(theme.default_theme),

    pub fn init() Toast {
        return .{};
    }

    pub fn show(self: *Toast, message: []const u8, toast_type: ToastType) void {
        self.message = message;
        self.toast_type = toast_type;
        self.visible = true;
        self.expires_at = std.time.milliTimestamp() + self.duration_ms;
    }

    pub fn showInfo(self: *Toast, message: []const u8) void {
        self.show(message, .info);
    }

    pub fn showSuccess(self: *Toast, message: []const u8) void {
        self.show(message, .success);
    }

    pub fn showWarning(self: *Toast, message: []const u8) void {
        self.show(message, .warning);
    }

    pub fn showError(self: *Toast, message: []const u8) void {
        self.show(message, .err);
    }

    pub fn hide(self: *Toast) void {
        self.visible = false;
    }

    pub fn isVisible(self: Toast) bool {
        if (!self.visible) return false;
        return std.time.milliTimestamp() < self.expires_at;
    }

    pub fn update(self: *Toast) void {
        if (self.visible and std.time.milliTimestamp() >= self.expires_at) {
            self.visible = false;
        }
    }

    pub fn render(self: Toast, writer: anytype) !void {
        if (!self.isVisible()) return;

        // Icon based on type
        const icon = switch (self.toast_type) {
            .info => "i",
            .success => "+",
            .warning => "!",
            .err => "x",
        };

        // Style based on type
        switch (self.toast_type) {
            .info => try self.style.secondary(writer),
            .success => try self.style.success(writer),
            .warning => try self.style.warning(writer),
            .err => try self.style.danger(writer),
        }

        try writer.writeAll("[");
        try writer.writeAll(icon);
        try writer.writeAll("] ");
        try writer.writeAll(self.message);
        try self.style.reset(writer);
    }

    pub fn renderCentered(self: Toast, writer: anytype, width: u16) !void {
        if (!self.isVisible()) return;

        const msg_len = self.message.len + 4; // "[x] " prefix
        const padding = if (width > msg_len) (width - @as(u16, @intCast(msg_len))) / 2 else 0;

        for (0..padding) |_| {
            try writer.writeAll(" ");
        }
        try self.render(writer);
    }
};

// Tests
test "Toast" {
    var toast = Toast.init();
    toast.showSuccess("Test message");

    try std.testing.expect(toast.visible);
    try std.testing.expectEqual(ToastType.success, toast.toast_type);
    try std.testing.expectEqualStrings("Test message", toast.message);
}
